handlers/papertrail/papertrail.go

```diff
@@ -26,50 +28,136 @@ type Config struct {
 	// Application settings
 	Hostname string // Hostname value
 	Tag      string // Tag value
+
+	// Advanced: tweak how we send logs to papertrail
+	// Capacity is the amount of log to queue before
+	// we start dropping, write timeout dictates how
+	// long we should wait before giving up on a write
+	Capacity       int
+	ConnectTimeout time.Duration
+	WriteTimeout   time.Duration
 }
 
 // Handler implementation.
 type Handler struct {
-	*Config
-
-	mu   sync.Mutex
+	url  string
+	c    *Config
 	conn net.Conn
+	q    *queue.Queue
 }
 
 // New handler.
 func New(config *Config) *Handler {
-	conn, err := net.Dial("udp", fmt.Sprintf("%s.papertrailapp.com:%d", config.Host, config.Port))
+
+	// defaults
+	// TODO: what should these be?
+	if config.Capacity == 0 {
+		config.Capacity = 20
+	}
+	if config.ConnectTimeout == 0 {
+		config.ConnectTimeout = 30 * time.Second
+	}
+	if config.WriteTimeout == 0 {
+		config.WriteTimeout = 5 * time.Second
+	}
+
+	// connect to papertrail
+	url := fmt.Sprintf("%s.papertrailapp.com:%d", config.Host, config.Port)
+	conn, err := connect(url, config.ConnectTimeout)
 	if err != nil {
-		panic(err)
+		// TODO: should we kill here?
+		stdlog.Printf("log/papertrail: couldn't connect to papertrail")
 	}
 
+	// TODO: see if papertrail can handle out of order
+	// logs. if so, we can make our write function
+	// thread-safe & adjust the concurrency here
+	q := queue.New(config.Capacity, 1)
```

Comment:
Ahh I was thinking we'd have a `Buffer` with the size/interval flushing. The channel stuff is basically a queue as-is but the flushing will definitely be common among them.

As far as dropping when the threshold is reached, yea I guess we just delegate and let the user deal with it ahah not sure, definitely a tricky situation hahah. `stdlog`'ing it is probably good at least, so they show up in local logs at least (or cloudwatch in the Lambda cas

Reply:
I suppose concurrency doesn't really matter either, if the Loggly and friends are rate-limiting or congested the concurrency wouldn't really help anyway.

Reply:
That's what I originally had, but after digging into TCP a bit more, it didn't really makes sense to buffer with an open TCP connection, TCP will already do that and probably do a better job than us.

Yep, the queue is basically just a channel, it just adds support for: concurrency, knowing when it's drained and dropping when it's full. For things like Firehose or Cloudwatch, you can issue HTTP requests concurrently, which would hopefully speed things up and if you have a really fast writer.

This TODO is just to see if we could have multiple goroutines writing on the TCP connection at once and Papertrail being able to rearrange the logs in the console based on the timestamp. Not too important though, since I think Papertrail isn't even granular enough for this to matter.

Oh good call on logging out when we're at capacity, I'm printing the error, but I should also write out the actual log. Anything else you'd like me to change / fix?

Reply:
Ohh for TCP yeah true, I was thinking for HTTP like Elasticsearch / Loggly etc where it's a lot more efficient to batch. The HTTP envelope alone is usually bigger than a log line so I think it's worth it there

Reply:
Ah, I see yah. I can add that original buffer back in then for us in some of the other providers? Ideally it would probably issue requests as fast as the reader can handle, but then start batching backed up logs. That might get tricky though haha.